src/Octoshift/AwsCredentialProvider.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using Amazon.Runtime;
+
+namespace OctoshiftCLI;
+
+public class AwsCredentialProvider
+{
+    private readonly AWSArgs _awsArgs;
+    public AwsCredentialProvider(AWSArgs awsArgs)
+    {
+        _awsArgs = awsArgs;
+    }
+
+    public AWSCredentials GetCredentials()
+    {
+        if (string.IsNullOrEmpty(_awsArgs.AwsCredentialType)) //load credentials from provider chain
+        {
+            AWSCredentials result = SessionCredentials();
+            if (result != null)
+            {
+                return result;
+            }
+            result = BasicCredentials();
+            if (result != null)
+            {
+                return result;
+            }
+            result = WebIdentityCredentials();
+            return result ?? throw new OctoshiftCliException("No AWS credentials could be found");
+        }
+        else
+        { //use specific way to load credentials
+            return _awsArgs.AwsCredentialType == "basic"
+                ? BasicCredentials()
+                : _awsArgs.AwsCredentialType == "session"
+                    ? SessionCredentials()
+                    : _awsArgs.AwsCredentialType == "webIdentity"
+                                    ? WebIdentityCredentials()
+                                    : throw new ArgumentException("Unknown AwsCredentialType:" + _awsArgs.AwsCredentialType);
+        }
+    }
+
+    private SessionAWSCredentials SessionCredentials()
+    {
+        var awsAccessKey = _awsArgs.AwsAccessKey;
+        var awsSecretKey = _awsArgs.AwsSecretKey;
+        var awsSessionToken = _awsArgs.AwsSessionToken;
+        return string.IsNullOrEmpty(awsAccessKey) || string.IsNullOrEmpty(awsSecretKey) || string.IsNullOrEmpty(awsSessionToken)
+            ? null
+            : new SessionAWSCredentials(awsAccessKey, awsSecretKey, awsSessionToken);
+    }
+
+    private AssumeRoleWithWebIdentityCredentials WebIdentityCredentials()
+    {
+        return string.IsNullOrEmpty(Environment.GetEnvironmentVariable("AWS_WEB_IDENTITY_TOKEN_FILE"))
```

Comment:
If you use this approach to authentication, do you also need to specify the role to assume? Or is that done before you use the `gh gei` CLI somehow? I'd love to understand the workflow a little better.

Reply:
You need to assume the role before you use the` gh gei` CLI.  
We are using the EKS cluster, we want to run GEI as a pod/job in our EKS cluster with a service account. The service account binds with an AWS IAM role. When we start the pod/job EKS will help us to assume the role and prepare the credentials files and env variables in the pod, so we don't need to assume it.